lapis/src/main/kotlin/org/genspectrum/lapis/model/VariantQueryCustomListener.kt

```diff
@@ -53,6 +70,191 @@ class VariantQueryCustomListener(
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid integer", e)
+                }
+            }
+
+            else -> {
+                throw BadRequestException("expression <= can not be used for ${field.type}", null)
+            }
+        }
+    }
+
+    override fun enterMetadataGreaterThanEqualQuery(ctx: VariantQueryParser.MetadataGreaterThanEqualQueryContext) {
+        val metadataName = ctx.geneOrName().text
+        val metadataValue = ctx.dateOrNumber().text
+
+        val field: SequenceFilterField? = allowedSequenceFilterFields.fields[metadataName.lowercase(Locale.US)]
+        field ?: throw BadRequestException("Metadata field $metadataName does not exist", null)
+        when (field.type) {
+            SequenceFilterFieldType.Date -> {
+                try {
+                    val date = LocalDate.parse(metadataValue)
+                    expressionStack.addLast(DateBetween(metadataName, to = date, from = null))
+                } catch (exception: DateTimeParseException) {
+                    throw BadRequestException("'$metadataValue' is not a valid date: ${exception.message}", exception)
+                }
+            }
+
+            is SequenceFilterFieldType.Float -> {
+                try {
+                    expressionStack.addLast(FloatBetween(metadataName, to = metadataValue.toDouble(), from = null))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid float", e)
+                }
+            }
+
+            is SequenceFilterFieldType.Int -> {
+                try {
+                    expressionStack.addLast(IntBetween(metadataName, to = metadataValue.toInt(), from = null))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid integer", e)
+                }
+            }
+
+            else -> {
+                throw BadRequestException("expression >= can not be used for ${field.type}", null)
+            }
+        }
+    }
+
+    override fun enterMetadataQuery(ctx: VariantQueryParser.MetadataQueryContext) {
+        val metadataName = ctx.geneOrName().text
+        val metadataValue = ctx.value().text
+
+        val field: SequenceFilterField? = allowedSequenceFilterFields.fields[metadataName.lowercase(Locale.US)]
+        field ?: throw BadRequestException("Metadata field $metadataName does not exist", null)
+        when (field.type) {
+            SequenceFilterFieldType.String -> {
+                expressionStack.addLast(StringEquals(metadataName, metadataValue))
+            }
+
+            SequenceFilterFieldType.Lineage -> {
+                val lineage = LineageEquals(metadataName, metadataValue, false)
+                expressionStack.addLast(lineage)
+            }
```

Comment:
I think we have to discuss how we get the `includeSublineages = true` case here. To me, it's currently not clear. The normal code does this:
https://github.com/GenSpectrum/LAPIS/blob/be1bb873aa0358f4d74eb69be3e08c913d445302/lapis/src/main/kotlin/org/genspectrum/lapis/model/SiloFilterExpressionMapper.kt#L282-L298
But it assumes that every lineage system uses this notation which is probably not true? For now, we could probably also use that logic here.

(Sidenote nitpick: Especially for boolean arguments I find the named arguments syntax really helpful)

Reply:
oh yes! thank you for mentioning this, I meant to ask and forgot @chaoran-chen and @fengelniederhammer should the generalized advanced queries include lineages? Will `*` always be used to specify the inclusion of sublineages? Should we add the handling of aliases or will SILO do this?

Reply:
I think for now it's probably good to assume that `*` means "including sublineages". But we should think about that separately again.

Handling of aliases can be left to SILO.

Reply:
I agree that we can assume a trailing `*` to mean including sublineages for lineage fields.